### PR TITLE
LLVM+newlib include path search fixes

### PR DESCRIPTION
**Pull request by gebart**
This PR contains some small fixes for the Makefiles to get the build system working with Clang/LLVM again after being broken since #6065.
Also included is a change to place C++ header includes before C includes on the command line when building C++ files. This was the original cause for the error about missing includes in #6065.
Added one more check for TARGET_ARCH in the LLVM makefile to avoid adding -target to CFLAGS when building for native.
Tested in a make buildtest with TOOLCHAIN=llvm set. Native and most Cortex-M platforms build successfully, stm32f2 and lpc2387 need some fixes in order to build with Clang. MSP430 and AVR are not supported by LLVM.

### CODE REVIEW

**Comment by cgundogan** [comment]
@tfar can you have a look at this one? IIRC, we talked about this last week?

**Comment by tfar** [comment]
Tried building hello world example with this branch on macOS with llvm/clang trunk with the following command line `env BOARD=samr21-xpro TOOLCHAIN=llvm make CFLAGS="--sysroot=/Users/tobiasm/Downloads/gcc-arm-none-eabi-5_4-2016q3 -isystem /Users/tobiasm/Downloads/gcc-arm-none-eabi-5_4-2016q3/includes"`.

AGAIK `--sysroot` is needed so it doesn't try to use the includes from `/usr/include…` and for unknown operating systems clang wants to know where to find the includes from the actual sysroot, so I've added the `-isystem…` stuff.

Got the following error:
```
tobiasm@moe ~/D/r/R/e/hello-world> env BOARD=samr21-xpro TOOLCHAIN=llvm make CFLAGS="--sysroot=/Users/tobiasm/Downloads/gcc-arm-none-eabi-5_4-2016q3 -isystem /Users/tobiasm/Downloads/gcc-arm-none-eabi-5_4-2016q3/includes"
Warning: no PORT set!
bash: arm-none-eabi-gcc: command not found
Building application "hello-world" for "samr21-xpro" with MCU "samd21".

/Users/tobiasm/Development/rep/RIOT/examples/hello-world/main.c:28:56: error: use of undeclared identifier 'RIOT_BOARD'
    printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
                                                       ^
/Users/tobiasm/Development/rep/RIOT/examples/hello-world/main.c:29:50: error: use of undeclared identifier 'RIOT_MCU'
    printf("This board features a(n) %s MCU.\n", RIOT_MCU);
                                                 ^
2 errors generated.
make[1]: *** [/Users/tobiasm/Development/rep/RIOT/examples/hello-world/bin/samr21-xpro/hello-world/main.o] Error 1
make: *** [all] Error 2
```

How do you build using clang for an embedded platform? What command line do you use?

**Comment by gebart** [comment]
The easiest solution is to add `/Users/tobiasm/Downloads/gcc-arm-none-eabi-5_4-2016q3` to `PATH` and not set any CFLAGS or anything else by hand. The Makefile should be able to figure it out if you have a cross GCC in PATH.

There are some things in the makefile that rely on finding a cross GCC in path, even when building with Clang. I'm guessing the error about arm-none-eabi-gcc not found comes from the following line:
https://github.com/gebart/RIOT/blob/e9f8608307bec7edc72caa17752bde8101af6434/cpu/Makefile.include.llvm#L43

You need to use `QUIET=0` to debug these kinds of errors.

If you don't want to add it to PATH then you could try setting `GCC_MULTI_DIR` manually. `GCC_MULTI_DIR` should be set to the output of `arm-none-eabi-gcc -print-multi-directory ...` with all your platform CFLAGS, in order to get the correct multilib directory for your particular CPU.
For me for the samr21 `GCC_MULTI_DIR` is set to `armv6-m`, but it depends on the multilib settings when building the toolchain (use `arm-none-eabi-gcc -print-multi-lib` to see the multilib specifications for your toolchain)

**Comment by tfar** [comment]
Right. The way you describe it works. However, if a user sets `TOOLCHAIN=llvm`, it feels a bit strange to still require a complete GCC cross compiling toolchain to be present in the PATH.

**Comment by gebart** [comment]
I agree that this is not optimal, but with this PR at least it builds.
Clang support in RIOT is experimental. Because of the way that LLVM is
designed, it is necessary to specify the system include paths when cross
compiling. The purpose of this PR is to provide a basic automatic
configuration for simple use cases. We also use the gcc provided libstdc++
for c++ support, so there's still some work to be done in order to build
RIOT without a gcc toolchain available at least.

File: cpu/Makefile.include.llvm [hunk 1]
```diff
@@ -31,19 +31,24 @@ export DBG         = $(GDBPREFIX)gdb
 GCC_CXX_INCLUDE_PATTERNS ?= \
   /etc/alternatives/gcc-$(TARGET_ARCH)-include/c++/*/ \
   /usr/$(TARGET_ARCH)/include/c++/*/ \
+  /usr/lib/gcc/$(TARGET_ARCH)/*/include/g++-v8 \
+  /usr/lib/gcc/$(TARGET_ARCH)/*/include/g++-v7 \
+  /usr/lib/gcc/$(TARGET_ARCH)/*/include/g++-v6 \
   /usr/lib/gcc/$(TARGET_ARCH)/*/include/g++-v5 \
```
Comment (OlegHahm) [inline cpu/Makefile.include.llvm]: Would `/usr/lib/gcc/$(TARGET_ARCH)/*/include/g++-v*` work, too?

**Comment by gebart** [comment]
The purpose is to test these paths in the order specified. Adding a
wildcard at the end makes it less obvious IMO and would require sorting
tweaks on the results.

File: cpu/Makefile.include.llvm (hunk 1 above)
Comment (OlegHahm) [inline cpu/Makefile.include.llvm]: Thanks for the explanation.

**Pull request merged**